### PR TITLE
abi: refactor abi package 

### PR DESCRIPTION
**Pull request by yoomee1313**
#### This PR is part of https://github.com/ethereum/go-ethereum/pull/18406 which supports tuple type. 
* It does not contain whole code. It only contains refactoring part. The another part will be contained in later.
* Tuple support has been already used in other parts of klaytn, but it is not applied in the official klaytn repo.

example solidity `nested.sol` - containing nested struct
```
pragma solidity ^0.4.24;

contract Nested {

    Parent public parent;
    struct Parent {
        uint a;
        Child child;
    }

    struct Child {
        uint b;
    }

}
```
abigen
```
$ abigen --sol nested.sol --pkg main --out nested.go
Failed to generate ABI binding: unsupported arg type: tuple
```

#### It refactors abi package code prior to apply tuple support.

https://github.com/klaytn/klaytn/commit/457866ee95ff03e8d8dbc174f87055876e8be7d5: Refactor code to support tuple support
* Made `ArgumentMarshalling` struct. It is used at `NewType` etc.
* Refactored `unpackTuple` and `unpackAtomic`. Acts exactly same as before.
* Renamed function `mapAbiToStructFields` to `mapArgNamesToStructFields`.
* Renamed function `getDynamicTypeOffset` to `getTypeSize`.
* Removed `getArraySize` and `getFullElemSize`. instead, replace those functions to `getTypeSize`.
* Trivial refactoring such as removing unused parameter and comment revision.

https://github.com/klaytn/klaytn/commit/4d69b6ece4fa28fa8f9d0df206e1bbe04926a9f8: Add components parameter to support tuple
* Added `ArgumentMarshalling` type parameter to the `NewType` function.
* Added `ArgumentMarshalling` variable to the `TestPack`, `TestTypeRegexp` tcs.
 
https://github.com/klaytn/klaytn/pull/792/commits/67e436bb5dbf4b2a9170cff0118131f382656894
* Add `components` argument to the `ArgumentMarshaling` struct.
* Remove unnecessary redundent testcase at `pack_test.go`

https://github.com/klaytn/klaytn/pull/792/commits/221460711cccb17149f88f4fb8f7d190a693473b
* add some test cases in abi_test.go 
* modify buff of TestMultiReturnWithStringSlice
## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: accounts/abi/pack_test.go [hunk 1]
```diff
@@ -33,303 +33,362 @@ import (
...
 			"bytes23",
+			nil,
 			[23]byte{1},
 			common.Hex2Bytes("0100000000000000000000000000000000000000000000000000000000000000"),
 		},
 		{
 			"bytes24",
+			nil,
 			[24]byte{1},
 			common.Hex2Bytes("0100000000000000000000000000000000000000000000000000000000000000"),
```
Comment (ehnuje) [inline accounts/abi/pack_test.go:303]: It seems that the elements starting from line 254 and line 259 look identical. And this seems to be the reason why deleted lines appeared in the original change. Could you please check?

File: accounts/abi/argument.go [hunk 2]
```diff
@@ -33,24 +33,26 @@ type Argument struct {
...
-	}
-	err := json.Unmarshal(data, &extarg)
+	var arg ArgumentMarshaling
+	err := json.Unmarshal(data, &arg)
 	if err != nil {
 		return fmt.Errorf("argument json err: %v", err)
 	}
 
-	argument.Type, err = NewType(extarg.Type)
+	argument.Type, err = NewType(arg.Type, nil)
```
Comment (ehnuje) [inline accounts/abi/argument.go:50]: Shouldn't it be like below?
```suggestion
	argument.Type, err = NewType(arg.Type, arg.Components)
```

File: accounts/abi/argument.go [hunk 3]
```diff
@@ -116,33 +140,38 @@ func (arguments Arguments) unpackTuple(v interface{}, marshalledValues []interfa
...
 				return fmt.Errorf("abi: insufficient number of arguments for unpack, want %d, got %d", len(arguments), value.Len())
 			}
 			v := value.Index(i)
-			if err := requireAssignable(v, reflectValue); err != nil {
+			if err := requireAssignable(v, reflect.ValueOf(marshalledValues[i])); err != nil {
 				return err
 			}
 
-			if err := set(v.Elem(), reflectValue, arg); err != nil {
+			if err := set(v.Elem(), reflect.ValueOf(marshalledValues[i])); err != nil {
```
Comment (ehnuje) [inline accounts/abi/argument.go:174]: Changing `set` to `unpack` will be applied later, right?

Comment (yoomee1313) [reply accounts/abi/argument.go:174]: @ehnuje Yes, it will be applied later on tuple support PR.

File: accounts/abi/pack_test.go (hunk 1 above)
Comment (yoomee1313) [reply accounts/abi/pack_test.go:303]: Thanks!! Applied at https://github.com/klaytn/klaytn/pull/792/commits/67e436bb5dbf4b2a9170cff0118131f382656894

File: accounts/abi/argument.go (hunk 2 above)
Comment (yoomee1313) [reply accounts/abi/argument.go]: Applied at https://github.com/klaytn/klaytn/pull/792/commits/67e436bb5dbf4b2a9170cff0118131f382656894

File: accounts/abi/abi_test.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ const jsondata2 = `
 ]`
```
Comment (aidan-kwon) [inline accounts/abi/abi_test.go:60]: Why don't you add test cases introduced in https://github.com/ethereum/go-ethereum/pull/18406 ?

Comment (yoomee1313) [reply accounts/abi/abi_test.go:63]: Thanks! Added https://github.com/klaytn/klaytn/pull/792/commits/221460711cccb17149f88f4fb8f7d190a693473b

**Review by ehnuje — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

File: accounts/abi/argument.go [hunk 5]
```diff
@@ -100,14 +102,37 @@ func (arguments Arguments) Unpack(v interface{}, data []byte) error {
...
+	if elem.Kind() == reflect.Struct {
+		fieldmap, err := mapArgNamesToStructFields([]string{argument.Name}, elem)
+		if err != nil {
+			return err
+		}
+		field := elem.FieldByName(fieldmap[argument.Name])
+		if !field.IsValid() {
+			return fmt.Errorf("abi: field %s can't be found in the given value", argument.Name)
+		}
+		return set(field, reflect.ValueOf(marshalledValues[0]))
```
Comment (KimKyungup) [inline accounts/abi/argument.go:122]: @yoomee1313 Will you change `set` to `unpack` in next PR?

**Review by KimKyungup — APPROVED** [review]
After merge all related code, we can compare this abi code with Geth.
Also, please consider to check if all abi test cases are tested.

**Pull request merged**